Export MIDI reflecting current in-app expression settings; numerous other changes

broadwell opened the pull request:
There's no particular rush to merge this -- the PR is partly just to get a convenient Netlify preview link.

The main contribution here is the in-app expression MIDI export functionality, though the PR also now includes several more updates that are nearly as consequential. Some no doubt could be moved to other PRs, but perhaps that's not necessary. The commit messages should (hopefully) explain all of the changes.

broadwell committed:
Comments to delineate stores groups

broadwell committed:
Wire in functions to export in-app expression MIDI

A couple of other minor improvements: 1) roll JSON metadata (title, composer, etc.) are passed to the SamplePlayer for eventual use when generating the in-app expression MIDI; 2) the location of the split between treble and bass keyboard pans is derived from the config data for each roll type.

broadwell committed:
Build in-app expression MIDI using current params

You get much better control over the final output MIDI from the ToneJS MIDI package if you assemble all of the data for it in a big JSON object and then load the JSON into a ToneJS MIDI instance. The actual API for building a MIDI instance on the fly, which is still used in the live-recording code (also in WebMidi.svelte), is much more limited. In this situation, the data comes from the roll metadata and the in-app expression notes, tempo and pedaling interval trees, which are computed at load time and whenever the in-app expression parameters are changed.

broadwell committed:
Some extra comments

broadwell committed:
Enable customization of soft pedal and accent mods

Why define these as arbitrary constants, when they can each have a slider? Note that these settings are best considered properties of the underlying piano/pianola hardware, rather than part of an expression emulation mechanism, which is why their sliders are in the Audio Settings menu along with the keyboard range velocity curves and sample playback settings, and why they aren't exported with the expression emulation settings.

broadwell committed:
Make VelocitySpliner fit panel without overflowing

This is various levels of cope due to the VelocitySpliner having inflexible container dimensions and also somehow always managing to overflow on Firefox for Mac at 100% width.

broadwell committed:
Apply trackerbar extension setting to pedal events for all roll types

broadwell committed:
Make button styles a bit more consistent

broadwell committed:
Add another piano playback knob: sustain linger

This is a bit experimental, but it helps to address the frequent observation that much of the sustain pedaling on reproducing rolls seems a bit too "dry" (abrupt) even given differing times and tastes, and so it might possibly be due to a shortcoming of the recording equipment. Note also a minor change to ensure that the play/pause icon reverts to the stopped state when playback reaches the end of the notes on the roll.

broadwell committed:
Roll navigator strip should behave properly after window resize

Just as changes in OSD 4 meant that the progress indicator overlay on the navigator strip wouldn't be drawn in the proper location after the browser window is resized (this was addressed in a previous PR), click and drag events on the strip after a window resize also don't work the way they used to in OSD 3. This change to the handler for both types of events seems to do the trick. Note that this commit probably could/should be a standalone branch and PR.

broadwell committed:
Don't list current roll in related works

This bug happened because the roll metadata from its JSON file confusingly didn't include the DRUID (which is in the filename). So we add it when the file is loaded.

broadwell committed:
Tweak wording of roll details pane

broadwell committed:
Swap out rewind icon

broadwell committed:
Make In-App/Expression MIDI toggle UI less clunky

broadwell committed:
No need to show both roll Label and Publisher in details pane

broadwell committed:
Display original composer when present in metadata

broadwell committed:
Provide contact info on the howto docs page

Open the spam floodgates...

broadwell committed:
Add docs re: latency, system specs and compatibility

broadwell committed:
Tweak docs navigation menu items

The "I want to..." landing page is already linked from "HOME" in the mock Spotlight header, so there's no need to include it in the documentation links. Also, "How to Play" is probably a better teaser text than "How to Use"

broadwell committed:
: after every changeable menu item (for consistency)

broadwell committed:
Enable tooltipAction on panel switcher buttons

This necessitated some changes to the styling and implementation of the panel switcher buttons, the most notable of which was making them IconButton components with custom styles. There's still a bit of a UX issue with the panel switcher buttons not having visible text labels (ironically this now more of an issue for visitors who aren't using screen readers); in particular it makes it awkward to specify which button to press when writing documentation. But at least the active tooltips are more immediately and visibly displayed when moused over than passive tooltips. Are such active tooltips also necessary for the transport control buttons? Maybe not, if only because their iconography is more conventional and readily understood.

broadwell committed:
Notification success green was a bit sickly

moritztim commented on src/components/RollDetails.svelte:
```diff
@@ -106,7 +106,7 @@
   </dd>
   {#if similarWorksByPerformer.length > 0}
-    <dt>Similar Works By This Performer</dt>
+    <dt>Other Works By This Performer</dt>
```
I see you chose the solution of changing the wording, rather than using a fancy matching algorythm to determine musical similarity :/

broadwell replied:
Yes, there's certainly a lot more we could do in the way of recommendations via basic metadata indexing or fancy roll similarity metrics, but for now it's best to be honest!

moritztim commented on src/Pianolatron.svelte:
```diff
@@ -247,6 +247,7 @@
     return Promise.all([mididataReady, metadataReady, pianoReady]).then(
       ([, metadataJson]) => {
         metadata = (({ holeData: _, ...obj }) => obj)(metadataJson);
+        metadata.druid = roll.druid;
```
That's a much simpler solution, I'll close [my pr](https://github.com/sul-cidr/pianolatron/pull/222).

broadwell committed:
Update $ticksPerSecond on tempo changes in in-app and MIDI expression

This fixes a bug whereby the "Roll Speed" readout in the ViewerMetrics component won't update on tempo adjustments during playback (which occur due to acceleration emulation) because the $ticksPerSecond store that triggers calculation of the Roll Speed for display is only updated on manual skip/stop/start events. During playback, the tempo setting was only changed in the MIDI player, which is not hooked reactively to the $ticksPerSecond store. A theoretically more elegant approach in which the MIDI player tempo and ticksPerSecond value are all linked reactively to a global $currentTempo store was also tried, but it was oddly not as performant and prone to glitches.

broadwell committed:
Buttons in recording notification were bleeding into close box

broadwell committed:
GitHub logo should appear to the right of the link.

Maybe there's a better way to do this?
